### PR TITLE
Add patch guideline documentation

### PR DESCRIPTION
**Pull request by MichaelSimons**
Related to https://github.com/dotnet/source-build/issues/2237.

### CODE REVIEW

**Review by omajid — APPROVED** [review]

File: Documentation/patching-guidelines.md [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+
+```sh
+git am "$PATCH_DIR/sdk/*"
+```
+
+This creates a Git commit with the patch contents, so you can easily amend a
+patch or create a new commit on top that you can be sure will apply cleanly.
+
+There is a method to create a series of patches based on a range of Git commits,
+but this is not usually useful for 6.0 main development. It is used in servicing
```
[the remark this answers is not in the document]
Comment (MichaelSimons) [reply Documentation/patching-guidelines.md]: I removed this paragraph as it is obsolete.  It was copied from https://github.com/dotnet/installer/tree/main/src/SourceBuild#creating-a-patch-file which I am replacing with a reference to this new documentation in https://github.com/dotnet/installer/pull/17127.

**Review by mmitche — APPROVED** [review]

File: Documentation/patching-guidelines.md [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+ time to flow into source build. This can block source build development
+(e.g. builds, tests, feature development, etc.). To avoid these delays source
+patches are applied within source-build to the repo source. These patches are intended
+to be ephemeral until the repo fixes flow in.
+
+This document provides guidance for creating and managing source build patches.
+
+## Creating Patches
+
+To create a repo patch file, first commit your changes to the repo as normal,
```
Comment (mthalman) [inline Documentation/patching-guidelines.md:16]:
```suggestion
To create a repo patch file, first commit your changes to the source repo (e.g. runtime, aspnetcore) as normal,
```

File: Documentation/patching-guidelines.md [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+then run this command inside the repo to generate a patch file inside the repo:
+
+```sh
+git format-patch --zero-commit --no-signature -1
+```
+
+Then, move the patch file into the [installer repo](https://github.com/dotnet/installer/tree/main/src/SourceBuild/patches)
+, at `src/SourceBuild/patches/<repo>`.
```
Comment (mthalman) [inline Documentation/patching-guidelines.md:24]:
```suggestion
, at `src/SourceBuild/patches/<repo>`. If an existing directory for the repo does not already exist, you will need to create one.
```

File: Documentation/patching-guidelines.md [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+To apply a patch, or multiple patches, use `git am` while inside the target
+repo. For example, to apply *all* `sdk` patches onto a fresh clone of the `sdk`
+repository that has already been checked out to the correct commit, use:
+
+```sh
+git am "$PATCH_DIR/sdk/*"
+```
+
+This creates a Git commit with the patch contents, so you can easily amend a
+patch or create a new commit on top that you can be sure will apply cleanly.
```
Comment (mthalman) [inline Documentation/patching-guidelines.md:42]: This should be in a separate section labeled as "Update an Existing Patch" and should have another paragraph at the beginning describing the scenario.

Comment (MichaelSimons) [reply Documentation/patching-guidelines.md:46]: I separated this into an `Applying Patches` section.  It is useful for ensuring patches apply, amending patches, and defining a new patch that touches code modified by existing patches.  The `Resolving Patch Conflicts` section is of course related.

**Review by mthalman — APPROVED** [review]

**Pull request merged**